src/VncScreenShare/vnc/FrameEncoder.cs

```diff
@@ -0,0 +1,66 @@
+﻿using VncScreenShare.Capture;
+using VncScreenShare.vnc.Rectangles;
+
+namespace VncScreenShare.Vnc
+{
+	internal class FrameEncoder
+	{
+		private readonly IEnumerable<ImgEncoding> m_supportedEncodings;
+		private readonly ZlibCompressor m_zlibCompressor;
+		private readonly HashSet<ImgEncoding> m_clientEncodings;
+
+		public FrameEncoder(IEnumerable<ImgEncoding> supportedEncodings)
+		{
+			m_supportedEncodings = supportedEncodings;
+			m_clientEncodings = new HashSet<ImgEncoding>();
+			m_zlibCompressor = new ZlibCompressor();
+		}
+
+		public ImgEncoding? ImageEncoding
+		{
+			get;
+			private set;
+		}
+
+
+		public void ClientEncodingsConfigured(IReadOnlyCollection<ImgEncoding> clientEncodings)
+		{
+			foreach (var encoding in clientEncodings)
+			{
+				if (encoding > 0)
+				{
+					m_clientEncodings.Add(encoding);
+				}
+			}
+		}
+
+		public Rectangle EncodeFrame(CapturedFrame capturedFrame)
+		{
+			if (ImageEncoding == null)
+			{
+				ChooseEncoding();
+			}
+
+			if(ImageEncoding == ImgEncoding.EncodingZlib) return new ZlibRectangle(capturedFrame, m_zlibCompressor);
+			return new RawRectangle(capturedFrame);
+		}
+
+		private void ChooseEncoding()
+		{
+			Console.WriteLine($"Client supports Encodings: {string.Join(";", m_clientEncodings)}");
+			Console.WriteLine($"Server supports Encodings: {string.Join(";", m_supportedEncodings)}");
+			ImageEncoding = ImgEncoding.RawEncoding;
+			foreach (var encoding in m_clientEncodings)
+			{
+				if (m_supportedEncodings.Contains(encoding))
+				{
+					ImageEncoding = encoding;
```

Comment:
With this simple algorithm,  the order of the lists is relevant. When the client sends "raw, zlib" as client formats, raw will be used,  because it's the first in the supported list.

Reply:
Maybe the client already sends an ordered list and the server should ideally comply with that ?

Reply:
Exactly the order is a hint wich one to choose
https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#setencodings